File: src/my_robot_controller/my_robot_controller/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounter(Node):
    def __init__(self):
        super().__init__('num_counter')
        self.num_count_publisher_ = self.create_publisher(
            Int64, 'number_counter', 10)
        self.num_subscriber_ = self.create_subscription(
            Int64, 'number', self.callback_number, 10)
        self.counter_ = 0
        self.service_ = self.create_service(
            SetBool, 'reset_number_counter', self.callback_reset_counter)
        self.get_logger().info("Number counter has been started!")

    def callback_number(self, msg):
        self.counter_ += msg.data
        next_msg = Int64()
        next_msg.data = self.counter_
        self.num_count_publisher_.publish(next_msg)
        self.get_logger().info(str(self.counter_))

    def callback_reset_counter(self, request, response):
        if request.data:
            self.counter_ = 0
            response.success = True
            response.message = "Successfully reset the counter!"
        else:
            response.success = False
            response.message = "Counter has not been reset!"
        return response


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounter()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/number_counter.cpp
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"


class NumberCounter : public rclcpp::Node {
public:
    NumberCounter() : Node("num_counter"), counter_(0) {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10,
            std::bind(&NumberCounter::callbackCounter, this, std::placeholders::_1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        server_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_number_counter",
            std::bind(&NumberCounter::callbackResetCounter,
                this, std::placeholders::_1, std::placeholders::_2));
        RCLCPP_INFO(this->get_logger(), "Number counter has been started!");
    }

private:
    void callbackCounter(const example_interfaces::msg::Int64::SharedPtr msg) {
        counter_ += static_cast<long>(msg->data);
        auto next_msg = example_interfaces::msg::Int64();
        next_msg.data = counter_;
        publisher_->publish(next_msg);
        RCLCPP_INFO(this->get_logger(), "%ld", counter_);
    }

    void callbackResetCounter(
        const example_interfaces::srv::SetBool::Request::SharedPtr request,
        const example_interfaces::srv::SetBool::Response::SharedPtr response) {
        if (request->data) {
            counter_ = 0;
            response->success = true;
            response->message = "Successfully reset the counter!";
        }
        else {
            response->success = false;
            response->message = "Counter has not been reset";
        }

    }

    long counter_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};

int main(int argc, char const* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
